### PR TITLE
A cleaner way to create Storage or Engine for tests

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

## What have you changed? (mandatory)

Provide `TempRocksStorageBuilder` and `TempRocksEngineBuilder`. No need to pass `ALL_CFS`, `TEMP_DIR` every where.

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

CI

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

**Comment by breeswish** [comment]
Also removed some `storage.stop()` in case of https://github.com/tikv/tikv/issues/3756

File: components/test_storage/src/sync_storage.rs [hunk 1]
```diff
@@ -31,7 +33,8 @@ pub struct SyncStorage<E: Engine> {
 
 impl SyncStorage<RocksEngine> {
     pub fn new(config: &Config, read_pool: ReadPool<storage::ReadPoolContext>) -> Self {
-        let storage = Storage::new(config, read_pool).unwrap();
+        let engine = TempRocksEngineBuilder::new().build().unwrap();
+        let storage = Storage::from_engine(engine, &config, read_pool).unwrap();
```
Comment (zhangjinpeng1987) [inline components/test_storage/src/sync_storage.rs]: s/&config/config

File: src/storage/mod.rs [hunk 2]
```diff
@@ -393,6 +392,63 @@ impl Options {
     }
 }
 
+/// A builder to build a temporary `Storage<RocksEngine>`.
+///
+/// Only be only used for test purpose.
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: Emmm, "Only used for..."?

File: components/test_storage/src/sync_storage.rs (hunk 1 above)
Comment (breeswish) [reply components/test_storage/src/sync_storage.rs]: Storage accepts `&Config`

File: src/storage/engine/rocksdb.rs [hunk 3]
```diff
@@ -125,6 +132,61 @@ impl Debug for RocksEngine {
     }
 }
 
+/// A builder to build a temporary `RocksEngine`.
+///
+/// Only be only used for test purpose.
+#[must_use]
+pub struct TempRocksEngineBuilder {
```
Comment (MyonKeminta) [inline src/storage/engine/rocksdb.rs]: Consider `TestEngineBuilder` or `LocalEngineBuilder` ?

File: components/test_storage/src/sync_storage.rs (hunk 1 above)
Comment (breeswish) [reply components/test_storage/src/sync_storage.rs]: Nice catch. Even clippy didn't find it.

File: src/storage/mod.rs [hunk 4]
```diff
@@ -393,6 +392,63 @@ impl Options {
...
+#[must_use]
+pub struct TempRocksStorageBuilder {
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: Consider `TestStorageBuilder` or `LocalStorageBuilder` ?

File: src/storage/mod.rs [hunk 5]
```diff
@@ -1566,17 +1604,16 @@ mod tests {
                 .async_get(Context::new(), Key::from_raw(b"x"), 101)
                 .wait(),
         );
-        storage.stop().unwrap();
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: Will it be properly stopped if we remove this line? I didn't understand clearly 🤔

Comment (breeswish) [reply src/storage/mod.rs]: It won't be properly stopped. However we will handle it in #3756

Comment (MyonKeminta) [reply src/storage/mod.rs]: How about remove these stops when it when it was done, but not in this PR?

Comment (breeswish) [reply src/storage/mod.rs]: I think both are ok, because there are already a lot of tests not closing it 🤣

**Pull request merged**